src/Lib.hs

```diff
@@ -0,0 +1,191 @@
+buildTriplet :: BinariesDB -> Triplet -> ExceptT String IO ByteString
+buildTriplet binDB trip@(Triplet bin pkg sys) = withBinariesDB binDB $ \conn -> do
+  errorFlags :: [Only Bool] <- liftIO $ query conn "SELECT error FROM binaries WHERE name = ? AND package = ? AND system = ?" (bin, pkg, show sys)
+  case errorFlags of
+    [Only True] -> do
+      bump conn
+      -- TODO
+      -- Right now, upon a cached failure, we can't tell the user any more than this.
+      -- Maybe we should store some more information about where the error came from, or a part of the nix log?
+      throwError "Known failure"
+    [] -> do
+      -- TODO thread safety issue: we have an unguarded insert with a primary key, this throws an exception if in the meantime someone else makes the same request
+      -- TODO maybe just use an atomic SQL transaction is enough?
+      res <- liftIO $ nixBuild trip
+      liftIO $
+        execute
+          conn
+          "INSERT INTO binaries (name, package, system, hits, error) VALUES (?,?,?,?,?)"
+          (bin, pkg, show sys, 1 :: Int, isLeft res)
+      ExceptT $ pure res -- equiv to either throwError pure
+    [Only False] -> do
+      -- TODO Don't invoke nix?
+      res <- liftIO $ nixBuild trip
+      case res of
+        Left _ -> error "what to do" -- TODO
+        Right bs -> bs <$ bump conn
+    _ : _ : _ -> error "impossible"
+  where
+    -- TODO is this the right way to do thread safety? or should we take more care on the application side?
+    bump conn = liftIO $ execute conn "UPDATE binaries SET hits = hits + 1 WHERE name = ? AND package = ? AND system = ?" (bin, pkg, show sys)
+
+nixBuild :: Triplet -> IO (Either String ByteString)
+nixBuild (Triplet bin pkg sys) = runExceptT $ do
+  -- TODO stream build process to stderr
+  -- TODO make sure we can't get any XSS-like shenanigans
+  (exit, stdout, stderr) <- liftIO $ Proc.readProcessWithExitCode "nix" ["build", "--no-link", "--print-out-paths", packageString ""] ""
+  liftIO . putStrLn $
+    unlines
+      [ "Finished a Nix build call",
+        "Exit code: " <> show exit,
+        "stdout: " <> stdout,
+        "stderr: " <> stderr
+      ]
+  case (exit, lines stdout) of
+    (ExitSuccess, [result]) -> do
+      let fullPath = result </> "bin" </> unBinaryName bin
+      liftIO . putStrLn $ "Successfully built " <> fullPath
+      fileExists <- liftIO $ Dir.doesFileExist fullPath
+      unless fileExists $ throwError "Package did not produce expected binary"
+      isExecutable <- liftIO $ Dir.executable <$> Dir.getPermissions fullPath
+      unless isExecutable $ throwError "Binary specified by the triplet was not an executable"
+      liftIO $ BS.readFile fullPath
+    _ -> throwError $ unlines ["An error occurred.", "  Exit code: " <> show exit, "  error:" <> stderr]
+  where
+    packageString = showString "nixpkgs#legacyPackages." . shows sys . showString ".pkgsStatic." . showString (unPackageName pkg)
+
+-- TODO what's a good name for this db?
```

Comment:
Typically, a web application does not have more than one sql database.
Following that practice, the name of db should be bound to the application rather than a table.
`ApplicationDB` or `BinplzDB` looks natural to me.

Reply:
Refactored into `ApplicationDB` in 8c08f4a76bca295c112791ef8d603d37676e88a7

Reply:
Okay then, `withBinariesDB` -> `withApplicationDB` as well?

Reply:
Ah, good point. 5e8ee57229cf08085edd67d6065d2420bf32850c